### PR TITLE
Add ModifyOrder api

### PR DESCRIPTION
**Pull request by DenisDeFaria**
Added the ModifyOrder Api and modified the example from order_and_cancel to order_and_modify_and_cancel

### CODE REVIEW

File: src/exchange/exchange_client.rs [hunk 1]
```diff
@@ -163,15 +164,32 @@ impl ExchangeClient {
         let timestamp = now_timestamp_ms();
         let vault_address = self.vault_address.unwrap_or_default();
 
-        let mut hashable_tuples = Vec::new();
+        // let mut hashable_tuples = Vec::new();
```
Comment (lmlmt) [inline src/exchange/exchange_client.rs:167]: nit: remove comment

File: src/exchange/exchange_client.rs [hunk 2]
```diff
@@ -163,15 +164,32 @@ impl ExchangeClient {
...
         let mut transformed_orders = Vec::new();
 
-        for order in orders {
-            hashable_tuples.push(order.create_hashable_tuple(&self.coin_to_asset)?);
-            transformed_orders.push(order.convert(&self.coin_to_asset)?);
-        }
+        let mut hashable_tuples_cloid = Vec::new();
+        let mut hashable_tuples = Vec::new();
+
+        match orders.iter().any(|order| order.cloid.is_none()) {
```
Comment (lmlmt) [inline src/exchange/exchange_client.rs:173]: nit: prefer if/else, e.g. `if orders.iter().any(|order| order.cloid.is_none() { ... } else { ... }`

File: src/exchange/exchange_client.rs [hunk 3]
```diff
@@ -163,15 +164,32 @@ impl ExchangeClient {
...
+                }
+            },
+            false => {
+                for order in orders {
+                    hashable_tuples_cloid.push(order.create_hashable_tuple_with_cloid(&self.coin_to_asset)?);
+                    transformed_orders.push(order.convert(&self.coin_to_asset)?);
+                }
+            }
+        };
+        let connection_id = match hashable_tuples_cloid.len() {
```
Comment (lmlmt) [inline src/exchange/exchange_client.rs:188]: nit: prefer if/else, e.g. `if hashable_tuples_cloid.is_empty() { ... } else { ... }`

File: src/exchange/exchange_client.rs [hunk 4]
```diff
@@ -183,6 +201,7 @@ impl ExchangeClient {
         self.post(action, signature, timestamp).await
     }
 
+
```
Comment (lmlmt) [inline src/exchange/exchange_client.rs:204]: nit: remove empty line

File: src/exchange/exchange_client.rs [hunk 5]
```diff
@@ -215,6 +234,7 @@ impl ExchangeClient {
         }
 
         let connection_id = keccak((hashable_tuples, vault_address, timestamp));
+        println!("connection_id: {:?}", connection_id);
```
Comment (lmlmt) [inline src/exchange/exchange_client.rs:237]: nit: remove println

File: src/exchange/order.rs [hunk 6]
```diff
@@ -96,6 +99,34 @@ impl ClientOrderRequest {
...
+
+    pub(crate) fn create_hashable_tuple_with_cloid(
+        &self,
+        coin_to_asset: &HashMap<String, u32>,
+    ) -> Result<(u32, bool, u64, u64, bool, u8, u64, [u8;16])> {
+        let hashed_order_type = self.order_type.get_type()?;
+        let &asset = coin_to_asset.get(&self.asset).ok_or(Error::AssetNotFound)?;
+        
+        // cloid is Some("0x1234567890abcdef1234567890abcdef".to_string()) is a 128 hex string
+        match &self.cloid {
```
Comment (lmlmt) [inline src/exchange/order.rs:113]: nit: prefer `if let Some(cloid) = &self.cloid { ... } else { ... }` for optional values

File: src/bin/order_and_modify_and_cancel.rs [hunk 7]
```diff
@@ -28,10 +28,11 @@ async fn main() {
         order_type: ClientOrder::Limit(ClientLimit {
             tif: "Gtc".to_string(),
         }),
+        cloid: Some("0x1234567890abcdef1234567890abcdef".to_string()),
     };
 
     let response = exchange_client.order(order, None).await.unwrap();
-    info!("Order placed: {response:?}");
+    println!("Order placed: {response:?}");
```
Comment (lmlmt) [inline src/bin/order_and_modify_and_cancel.rs:35]: why println?

**Review by lmlmt — COMMENTED** [review]
Thanks for putting this up! Couple of comments before merging

**Comment by DenisDeFaria** [comment]
Added bulk modify api

**Comment by lmlmt** [comment]
@DenisDeFaria this PR seems to have caused some issues for other users of the API, i.e. cancels are broken now. Feel free to test out these changes on your end and resubmit a new PR when it's ready.

**Comment by DenisDeFaria** [comment]
> @DenisDeFaria this PR seems to have caused some issues for other users of the API, i.e. cancels are broken now. Feel free to test out these changes on your end and resubmit a new PR when it's ready.

Could you give more details about the error, so I can quickly fix it, and if by any chance you have a way of reproducing it ?

Thanks

**Comment by Nayshins** [comment]
@lmlmt @DenisDeFaria I've been running this locally but have not seen any issues with canceling orders. Do you have any direct logs where this is occurring?

**Pull request merged**